src/app/SecurityPanel/Credentials/CreateCredentialModal.tsx

```diff
@@ -0,0 +1,227 @@
+  progressChange?: (inProgress: boolean) => void;
+}
+
+export const AuthForm: React.FC<AuthFormProps> = ({ onDismiss, onPropsSave, progressChange, ...props }) => {
+  const context = React.useContext(ServiceContext);
+  const addSubscription = useSubscriptions();
+  const matchExprService = React.useContext(SearchExprServiceContext);
+  const [matchExpression, setMatchExpression] = React.useState('');
+  const [matchExpressionValid, setMatchExpressionValid] = React.useState(ValidatedOptions.default);
+  const [loading, setLoading] = React.useState(false);
+
+  const [targets, setTargets] = React.useState<Target[]>([]);
+
+  const onSave = React.useCallback(
+    (username: string, password: string) => {
+      setLoading(true);
+      addSubscription(
+        context.api.postCredentials(matchExpression, username, password).subscribe((ok) => {
+          setLoading(false);
+          if (ok) {
+            onPropsSave();
+          }
+        })
+      );
+    },
+    [addSubscription, onPropsSave, context.api, matchExpression, setLoading]
+  );
+
+  React.useEffect(() => {
+    progressChange && progressChange(loading);
+  }, [loading, progressChange]);
+
+  React.useEffect(() => {
+    addSubscription(context.targets.targets().subscribe(setTargets));
+  }, [addSubscription, context.targets, setTargets]);
+
+  React.useEffect(() => {
+    let validation: ValidatedOptions = ValidatedOptions.default;
+    if (matchExpression !== '' && targets.length > 0) {
+      try {
+        const atLeastOne = targets.some((t) => evaluateTargetWithExpr(t, matchExpression));
+        validation = atLeastOne ? ValidatedOptions.success : ValidatedOptions.warning;
+      } catch (err) {
+        validation = ValidatedOptions.error;
+      }
+    }
+    setMatchExpressionValid(validation);
+  }, [matchExpression, targets, setMatchExpressionValid]);
+
+  return (
+    <CredentialAuthForm {...props} onSave={onSave} onDismiss={onDismiss} focus={false} loading={loading}>
+      <FormGroup
+        label="Match Expression"
+        labelIcon={
+          <Popover
+            headerContent="Match Expression Hint"
+            bodyContent={
+              <>
+                Try an expression like:
+                <MatchExpressionHint target={targets[0]} />
+              </>
+            }
+            hasAutoWidth
+          >
+            <Button
+              variant="plain"
+              aria-label="More info for match expression field"
+              onClick={(e) => e.preventDefault()}
+              className="pf-c-form__group-label-help"
+            >
+              <HelpIcon />
+            </Button>
+          </Popover>
+        }
+        isRequired
+        fieldId="match-expression"
+        helperText={
+          matchExpressionValid === ValidatedOptions.warning
+            ? `Warning: Match expression matches no targets.`
+            : `
+        Enter a match expression. This is a Java-like code snippet that is evaluated against each target
```

Comment:
Would it make sense to change Java-like to JavaScript like? Since the match expression evaluation is actually evaluating js code.

Reply:
The fact that it's JavaScript is sort of an internal implementation detail that might change in the (distant) future. This should really be a much more limited scripting language that can't do all of what the JS engine is capable of doing, which is why I had to lock down the JS parser so much on the backend when it validates expressions before evaluating them. I would rather have done it so that this uses an actual subset of Java syntax directly rather than running it through a JS engine, but it was a pragmatic choice to get something running in reasonable fashion, and not something I want to say we're committing to.

Reply:
Makes sense! I will keep the Java for now then :D

Reply:
Another possible avenue, rather than implementing a subset of Java syntax as an embedded scripting language, is to use JShell or something similar as a replacement for the current nashorn JS engine. If there is a way to similarly lock down the JShell parser so that it rejects expressions with dangerous constructs then this would mean the expression syntax really is Java, instead of Java-like.

https://github.com/cryostatio/cryostat/issues/990